[WIP] Differential build for each PR

A-Hilaly opened the pull request:
Discussion made before in #136

ryukinix commented:
![image](https://user-images.githubusercontent.com/7642878/32694212-3bbf64c2-c719-11e7-80ae-4d7b53f38c99.png)

There is something that I can do to help you, @A-Hilaly? Since staged files only occurs on pre-commit stage of development, this makes little sense to do here.

What are you  trying to accomplish with staged mode on travis? (sorry If I'm misunderstanding your job here)

A-Hilaly commented:
@ryukinix I was just trying to make the pipe thing work ( I never did before ! ) I am sorry for the mess.
I am actually wondering what `python stats.py --search `should receive exactly in ./test : files that were modified and added (then python should treat em to find what to build exactly) , or treat them with bash script and give `python stats.py --search `problem directories and their languages

ryukinix commented:
You can use `xargs` to send the output of one commad as arguments from `pipe` instead reading the STDIN.  Something like that: `diff_command | xargs python stats.py ...`. Let's say that `diff_command` return path1 path2 path2 as output so the `xargs python stats.py` will be called that way: `python stats.py path1 path2 path3`.

I know that you are only doing tests, but since you can came on this to make a proper solution, I would prefer that you don't break the `stats.py` usage of `--search` argument. You can implement another flag like `--files` or `--solutions` and call as `diff_command | xargs python3 --files`.

That you can enable by default the `--build` option, because doesn't makes sense using `--files` argument without building the solutions, right? Sorry for cannot help you a lot on this now since this code stats.py is mine (busy at college stuff :cry: )

ryukinix commented:
But feel free to try whatever approach you like and we can review later, ok?

Thanks for trying to implement #136 ! :)

A-Hilaly commented:
> this makes little sense to do here
here is a cleaner output of what I wanted to see
<img width="329" alt="args" src="https://user-images.githubusercontent.com/10897901/32694367-456574ec-c73e-11e7-9f3c-147df32123aa.png">

ryukinix commented:
The `--search` flag for default only receive a unique string designed to be a "language" name supported  for search for solution no the file tree of the repository. This flag doesn't take care if you try pass filename/filepath of solutions or another things. Sorry for this tricky and non-intuitive usage, I didn't think a lot what I wrote this.

ryukinix commented:
My advise is that you don't rely on the `--search` for do this task since this is completely unrelated. You'll need to combine the another functions provided on `stats.py` to create another flag, as I suggested, a `--files` flag, which can receive a list of parameters separated by spaces.

ryukinix commented:
https://github.com/DestructHub/ProjectEuler/blob/8b2e23b354c47235d9d10d0003bb33bc983ecef2/stats.py#L594
https://github.com/DestructHub/ProjectEuler/blob/8b2e23b354c47235d9d10d0003bb33bc983ecef2/stats.py#L321

Take a look these two lines for see what `options.search` is expected to be.

A-Hilaly commented:
> You can use xargs

Thank you ! I never head about it 🤦‍♂️ sure it can solve the problem here I m gonna try to implemented it

> You can implement another flag like --files or --solutions

I hesitated before doing it since thats what was discussed in #136 
it true that it's better than complexing --search option

A-Hilaly commented:
@ryukinix I agree that adding a new option is the best way here. But what do we need to pipe exactly into `./stats.py --build --search` direct out of git diff commands and then python should treat this or doing it in shell before so that would be easy for stats.exs also

ryukinix commented:
> Thank you ! I never head about it 🤦‍♂️ sure it can solve the problem here I m gonna try to implemented it

You are fine. Take your time and figure out what you think is best, but I think `xargs` is a good idea here.

> I hesitated before doing it since thats what was discussed in #136
it true that it's better than complexing --search option

I understand that, but in that case will be better not breaking old usage of this script since they have another meaning. The `--search` option is easy to use that you want to do a comparison of all solutions between two or more language (or even just one): `stats.py --search python --search haskell --build`. So I don't want break that.

> @ryukinix I agree that adding a new option is the best way here. But what do we need to pipe exactly into ./stats.py --build --search direct out of git diff commands and then python should treat this or doing it in shell before so that would be easy for stats.exs also

We can do the same with `stats.py --build --files`. If is so annoying to create another flag, we can even just forget about `--search` or `--files` and if no one flags is provided, read from stdin and build! HAHA
That way will be:

`diff_commands | stats.py` 

More simple than that is impossible.

A-Hilaly commented:
> Take a look these two lines for see what options.search is expected to be.

if I understand more now; ./stats.py --build --files should just change `query` and `language` with maybe a list of problems to build 

> We can do the same with stats.py --build --files

I meant stats.py --build --files ! My question was a bit shy 

> That way will be: 
`diff_commands | stats.py`
More simple than that is impossible.

Reallly ?!!  XD

A-Hilaly commented:
I will try to complete to this PR ASAP and ping you when things get clearer !

ryukinix commented:
@A-Hilaly thanks! :) so ping me later.

A-Hilaly commented:
<img width="512" alt="travispe" src="https://user-images.githubusercontent.com/10897901/32704705-8cd90cb4-c809-11e7-8baa-824feb4b5f8d.png">

@ryukinix Here is what actual output looks like with new problems 300, 301 (examples) and problem 501 modification.

A-Hilaly commented on stats.py:
```diff
@@ -201,6 +201,22 @@ def execute(self):
 # Cmdline parsing definitions
 # #
 
+def _callback(option, opt_str, value, parser):
```
this doesn't accept pipe anymore but make parsing a list of arguments easier. Is there a better way ?

A-Hilaly commented on stats.py:
```diff
@@ -390,7 +414,7 @@ def load_dataframe():
 
 
-def solutions_paths(df):
+def solutions_paths(df, from_files=None):
```
im not sure about implementing this here, maybe I am repeating some code out side ?

ryukinix replied:
well, this is a little hacky, but since this change little code we can accept that for now

ryukinix replied:
but this really should not be here.... probably we'll need a full refactor of `stats.py` later, probably neither I understand all this script although I wrote all this shit.. sorry for you have to working with this crazy build system hahaha

A-Hilaly replied:
hahaha x), its routine don't worry :p

A-Hilaly replied:
> probably we'll need a full refactor of stats.py later

I agree on that, some changes and classes maybe make things clearer and easier de fix

ryukinix replied:
Using functions is ok, the main problem is rely on `pandas` to do a lot of operations. But probably I will not remove that because I need pandas these days to generate some statistics stuff (like graphs) which this was the main reason to write this script (not only for testing/building), so doesn't makes sense remove that. But indeed I need write in a more sane way.

A-Hilaly commented on stats.py:
```diff
@@ -437,6 +464,17 @@ def count_solutions(df, solutions=True):
 
     return df_
 
+def handle_files(files):
```
same here

ryukinix replied:
why this docstring empty? this functions seems so hacky... why are you counting the `/` anyway?

A-Hilaly replied:
double "/" for "ProblemXXX/Lang/solution
one "/" is for readme and hashes so its ignored
none of them means a file in /ProjectEuler directory

ryukinix replied:
ok

A-Hilaly commented on stats.py:
```diff
@@ -590,8 +629,17 @@ def handle_options(options):
     query = [x.lower() for x in options.search]
+    unstaged_solutions = []
+    unstaged_core_files = []
+    tbsolutions = []
```
unstaged_solutions for new / modified solutions
unstaged_core_files for new/ modified core files ( stats.py, test, stats.exs ) 

unstaged_core_files the only utility I see for this is: stats.py changes would force stats.py --build to build all problems ignoring all other new/adds. I think decision making should be done in ./test script.

ryukinix replied:
This seems ok to me.

ryukinix commented on Problem206/Haskell/solution_1.hs:
```diff
@@ -6,6 +6,10 @@ where each “_” is a single digit.
 
 {--
+New comment
+--}
```
why this?

ryukinix commented on Problem300/Python/solution_1.py:
```diff
@@ -0,0 +1,3 @@
+
+if __name__ == "__main__":
```
and this? only testing?

A-Hilaly replied:
Yes I commented problem 501 and added 300 / 301 for testing

ryukinix replied:
ok

ryukinix commented on Problem501/Python/solution_slow_1.py:
```diff
@@ -33,4 +33,7 @@ def f_p(n):
     return last
-f_p(10 ** 12)
\ No newline at end of file
+
```
this two

ryukinix commented on stats.py:
```diff
@@ -208,17 +208,13 @@ def _callback(option, opt_str, value, parser):
-            if arg[:1] == "-" and len(arg) > 1:
-                break
-            value.append(arg)
+    for arg in parser.rargs:
```
why we need that?

ryukinix commented on stats.py:
```diff
@@ -504,7 +523,7 @@ def execute_builder(b):
 
 # need docs
-def build_result(df, ignore_errors=False, blame=False):
+def build_result(df, ignore_errors=False, blame=False, only=[]):
```
Using mutable structures (like lists) as default parameters is evil, consider to review that. You can use tuples.

A-Hilaly replied:
Thats true, maybe just none and assert its iterable after calling ?

ryukinix replied:
Yes, you can use None too. But probably tuple is better since you expect to receive a iterable object.

A-Hilaly replied:
Alright !

ryukinix commented:
> @ryukinix Here is what actual output looks like with new problems 300, 301 (examples) and problem 501 modification.

This is example is from your staged files on your local repo or travis? The result seems great, but I'm not sure about `./test --staged-only` flag if this example were ran on travis. If this is your output on your local machine this is completely ok.

ryukinix commented on test:
```diff
@@ -0,0 +1,19 @@
+
+cmd="$1"
+
+if [ "$cmd" = "--staged-only" ] || [ "$cmd" = "-so" ]; then
```
`--staged-only` means the `stage` files before commit them, so this doesn't make sense on running on travis.

This type of usage is only designed to use on local changes. 

I suggest you to change this flag for now as `--commited-only` or something more clear, `--from-master`, `--master`. But `--staged-only` is [another thing](https://softwareengineering.stackexchange.com/questions/119782/what-does-stage-mean-in-git), ok?

A-Hilaly replied:
> But --staged-only is another thing, ok? 

Yes , I will rename almost everything to make things clearer. Btw how many commands to we need at /test file : --unstanged, --master maybe ?

ryukinix replied:
I think only `--staged` and `--master` is enough. And later `--sync` and `--async` to select the desired implementation, but this can be in another PR since we don't have yet a proper implementation of this on `stats.exs`.

ryukinix commented on .travis.yml:
```diff
@@ -1,7 +1,10 @@
 
 script:
-  - docker-compose run async --quiet
+  - sudo pip install -r requirements.txt
```
We need to implement a new `docker-compose` task like `diff-sync`. We should not install this requirement on PR because this is pretty slow and we already have the docker image with all that working fine.

I'm ok to merge that in this way if you can open a issue talking about that: "Create a new service on docker-compose for diff-sync". Or something like that.

A-Hilaly replied:
Agreed

ryukinix replied:
I meant that: https://github.com/DestructHub/ProjectEuler/blob/8b2e23b354c47235d9d10d0003bb33bc983ecef2/docker-compose.yml#L3

ryukinix commented on stats.py:
```diff
@@ -218,7 +234,15 @@ def execute(self):
+
+parser.add_option(
+    "-f", "--files",
+    help="Choose the languages for print information",
```
This help text is a little confuse. You can rewrite it?

A-Hilaly commented:
at travis build ##623 I got this error
<img width="989" alt="errortrav" src="https://user-images.githubusercontent.com/10897901/32725809-d4d0e39e-c876-11e7-93b9-0e4d42d0a2ad.png">

I have doubt that I m missing something maybe full it need a full directory ?

ryukinix commented on test:
```diff
@@ -6,14 +6,33 @@ function diff_files () {
+if [ "$cmd" = "--commited-only" ] || [ "$cmd" = "-c" ]; then
+    build_commited
+elif [ "$cmd" = "--all" ] || [ "$cmd" = "-a" ]; then
+    echo "heheheh"
```
LOL

A-Hilaly replied:
xd you found it I tried to delete them all x)

ryukinix commented on .travis.yml:
```diff
@@ -1,7 +1,7 @@
 
 script:
-  - docker-compose run async --quiet
+  - docker-compose run diff-sync
```
Now this ok

A-Hilaly replied:
Actually when I made last changes test.sh doesn't find `git` command
<img width="762" alt="errt" src="https://user-images.githubusercontent.com/10897901/32730592-e2e9ca66-c887-11e7-8154-acf3ac0d2248.png">

ryukinix replied:
hmmmm, yes, I understand... I need to push your changes of `Dockerfile` to the dockerhub...

ryukinix commented on test.sh:
```diff
@@ -0,0 +1,35 @@
+    opt="$1"
+    mode="$2" # sync/async
+    if [ "$opt" = "--master-only" ] || [ "$opt" = "-m" ]; then
+        git checkout master
```
why we have `--master-only` and `--master-with-commited`?

I think don't makes sense doing `git checkout` here... this can be tricky. You mean `--master-only` to only build the master? You really can do that easily: checkout in the desired branch and using that script... I think we should insert a sensitive git operation inside of this script.

I know that probably things being messy in our discussion, but we should review that (only this part now `./test`). We have another problem here too because the order of arguments matters (the ideal is that doesn't matters, to make things more simple)

Let me be clear, I suggest this CLI design:

* `./test`: build sync and use the diff from master
* `./test sync`: build sync and use the diff from master
* `./test async`: build async and use the diff from master
* `./test [sync|async] diff`: use the diff from master
* `./test [sync|async] staged`: use the diff of staged files (not commited yet) from `git diff --cached --name-only`

And we need a little help function to be printed for `-h` or `--help` flags.

ryukinix replied:
If you think we have a better option, we need discuss it. I don't like very much the usage of the current implementation. We need something intuitive what the command do only seeing the command line:

How I can differ `--master-only`, `--commited-only` and `--master-with-commited` without looking to the source? I think this is not clear enough.

A-Hilaly replied:
> why we have --master-only and --master-with-commited?
You mean --master-only to only build the master?

Yes I wasn't focused when writing that ... useless feature

Things are much more clearer now ! ill make the changes now.
One more questions to complete the schema in my head for `./test [sync|async] staged`:
 - can  `git diff --cached --name-only` show different results than `git diff --name-only`?  In PR case ? 
- does `git diff --cached --name-only`  works only if we use `git add [files] ` ?
- are commits in PR considered as stage files for the main repo ?

PS : I already tried to see `git diff --cached --name-only` , output is empty

A-Hilaly replied:
Mmmm maybe it is just for users before even committing in their own fork ?

ryukinix replied:
> can git diff --cached --name-only show different results than git diff --name-only? In PR case ?

This is designed always to be executed only on local changes. We have not any result on PR travis system because we don't have `staged` files. If you are not very familiar with git, the `staged` files are after `git add` and before `git commit`. We have only the files after commit them on PR build.

> does git diff --cached --name-only works only if we use git add [files] ?

Yes.

> are commits in PR considered as stage files for the main repo ?

No. Commits on PR is unrelated to stage files, as I said on the first answer, stage files only exists after `git add` and before `git commit`. It's something like a `pre-commit` stage.

> PS : I already tried to see git diff --cached --name-only , output is empty

This is expected if you already commited the changes. The design to use `staged` test is to test your solution before commit them, that way avoiding pain and suffering to create more useless commits to fix sometimes something trivial.

ryukinix replied:
> Mmmm maybe it is good for users before even committing in their own fork ?

I don't have sure if I understood that correctly, but I think that users SHOULD test your commits or stage files in own fork.

A-Hilaly replied:
Thank you for answers ! I have better understanding of git diff manipulations.

> I don't have sure if I understood that correctly, but I think that users SHOULD test your commits or stage files in your own fork.

Yes exactly ! Users can be test answers in their own machine before even committing in their fork.

A-Hilaly commented:
Well need to test all the commands before merging also remove problem 300 and 301 and comments on 501

ryukinix commented:
> I willl need to test all the commands before merging also remove problem 300 and 301 and comments on 501 ! Just after test result

I'm building a new docker image after your edition of Dockerfile. Since I send it, you can test on travis too. This can be a too slow because you naively changed the Dockerfile near on top, so I need rebuild a lot of things. The ideal of working with Dockerfiles if you need modify, ensure to put new statements near of bottom because doing that you can cache the earlier layers (each command is a layer on Dockerfile).

A-Hilaly commented:
> The ideal of working with Dockerfiles if you need modify, ensure to put new statements near of bottom because doing that you can cache the earlier layers (each command is a layer on Dockerfile).

I understand what you mean. Shall I just put my changes on the bottom of the file ? it will not change many things I guess. 
PS: In line 13/14 in my dockerfile version I chmod and add test.sh ( changed `./test `to `./test.sh`) because it wasn't accepting it without extension . Do we really need `ADD ./test.sh . `there ?

ryukinix commented:
> I understand what you mean. Shall I just put my changes on the bottom of the file ? it will not change many things I guess. 

In the next time yes, for now keep in that way since I already build the image. If you change again you'll remove and I'll need to build again...

> PS: In line 13/14 in my dockerfile version I chmod and add test.sh ( changed ./testto ./test.sh) because it wasn't accepting it without extension . Do we really need ADD ./test.sh .there ?

Yes, we need the `ADD` stuff. I don't know about the extension problem, I never saw that before.

A-Hilaly commented:
> In the next time yes, for now keep in that way since I already build the image. If you change again you'll remove the current changes and I'll need to build again...

Painful x'(, I will try to not do that again !

> EDIT: actually I think we don't need that...

" I will try to not do that again !"

ryukinix commented:
@A-Hilaly it's ok, if you are learning that is ok. No one must have know everything before... knowing it.

I'm editing the Dockerfile and will release a simple commit fixing that (I'm building a new image again). I need that because after build, with is already very slow, I need to push it... my internet connection is not very good and a completly new docker build (because we edit near on top) is almost 500MB (yes, just for adding the git package XD).

But putting just `RUN apt-get install git -y` on bottom of file we can fix that easilly. That way we should as well renaming again `test.sh` to `test` to keep more consistent as `add` is without extension (the bash scripts of repo).

ryukinix commented on test.sh:
```diff
@@ -0,0 +1,46 @@
+mode=$1
+target=$2
+
+if [ "$mode" = "" ]; then
```
really better

ryukinix commented on test.sh:
```diff
@@ -0,0 +1,46 @@
+    status=1
+elif [ "$mode" = "sync" ]; then
+    if [ "$target" = "diff" ]; then
+        build_uncommited
```
why `build_uncommited`? this actually build the commited changes :v

ryukinix commented:
@A-Hilaly I made some changes, please `git pull --rebase` on your local copy repo before  push new things, ok?

ryukinix commented:
I'm pushing the new docker image to `destructhub/project_euler` and until night we should have a new image online with `git`.

ryukinix commented:
This will really take some time...

A-Hilaly commented:
> No one must have know everything before... knowing it.

👍 

> @A-Hilaly I made some changes

Thank you !

> I need that because after build, with is already very slow

I can try with my internet connection if you want I just need what to type and how do I send you the image (maybe dropbox), tell me ASAP so I can run it 

> and until night we should have a new image online.

its 20:34 here x) whats I guess we are 5/4 hours difference I guess

A-Hilaly commented:
@ryukinix I can also try to make it with a amazon instance they must have good internet connection :D

ryukinix commented:
@A-Hilaly that will be awesome, since `destructhub` on dockerhub is a organization I can easilly add you with write permissions :) (now here is really hard to upload that)

Earlier, the first docker images for destructhub, I used a VPS on DigitalOcean, but I don't have more credit :/

Please, point me by email or even here your dockerhub account then I'll add to the dockerhub/destructhub organization too :D

ryukinix commented:
@A-Hilaly what you need do with docker (after you have a properly installation of docker) on root of the repository:

```
docker login 
docker build -t destructhub/project_euler . && docker push destructhub/project_euler
```

And all be fine, but before that I need your dockerhub credentials (if you don't have you'll need create on dockerhub.io)

A-Hilaly commented:
I just created one ( login : ahilaly) ill prepare an instance and run it on screen mode all the night

ryukinix commented:
On Mon, Nov 13, 2017 at 5:07 PM, M.A.H. <notifications@github.com> 
wrote:
> ahilaly

I added you right now to destructhub organization. Try to see later if 
you received some email about confirmation.  I added as well no your 
write rights for project_euler docker image. Ping me later if you have 
some problem in pushing it.
-- 
As lembranças não são só do passado. Podem ser de agora e até de 
amanhã.
— Serial Experiments Lain
Manoel Vilela,
Discente em Engenharia da Computação,
Universidade Federal do Ceará.

A-Hilaly commented:
I started it in my Mac since I have everything required, on amazon machine it keeps failing after logging with this line:
`Cannot connect to the Docker daemon. Is the docker daemon running on this host?`

A-Hilaly commented:
> I added you right now to destructhub organization. Try to see later if
you received some email about confirmation.  I added as well no your
write rights for project_euler docker image. Ping me later if you have
some problem in pushing it.

Yes i received organization add confirmation.

btw i see a lot of warnings with the build is it fine 🤔 ?

ryukinix commented:
Yes, this is pretty normal. HAHAHA @lubien likes this behavior for 
instance
-- 
As lembranças não são só do passado. Podem ser de agora e até de 
amanhã.
— Serial Experiments Lain
Manoel Vilela,
Discente em Engenharia da Computação,
Universidade Federal do Ceará.

On Mon, Nov 13, 2017 at 5:18 PM, M.A.H. <notifications@github.com> 
wrote:
>> I added you right now to destructhub organization. Try to see later 
>> if
>> you received some email about confirmation. I added as well no your
>> write rights for project_euler docker image. Ping me later if you 
>> have
>> some problem in pushing it.
>> 
> Yes i received organization add confirmation.
> 
> btw i see a lot of warnings with the build is it fine 🤔 ?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

A-Hilaly commented:
XD alright then, things are going pretty good I expect it finishes in 1 or 2 hours, seems like docker hub are them selfs limiting download speed (from their side) to 128kb/s.

ryukinix commented:
@A-Hilaly since you finish that, can you restart the last build on travis?

A-Hilaly commented:
@ryukinix it is still pushing its too much data ;D

ryukinix commented:
Today hub.docker.com seems unstable. I have some difficulties to use the platform (for add you to in the organization).

ryukinix commented:
But we're close to finish that :)

ryukinix commented:
Since we got this working, would be really nice for new comers seeing your performance. It would be nice later if you can do something to add the output of the build (stats.py) as comment of the PR. That way we can discuss about the performance of the solution, if the answer is right and etc. What do you think? We can do that later, I think can be a good thing :) (since is painful looking for travis build always even when works)

A-Hilaly commented:
Very nice idea 👍 getting results as comment least it. will avoid going to travis every time to check results stats. well need an account for the organization ! DestructHub Bot ? 
( if we will not be obliged to create another docker image xd ) to use [PyGithub](http://chase-seibert.github.io/blog/2016/07/22/pygithub-examples.html) , or just normal requests with GitHub api ?

A-Hilaly commented:
<img width="588" alt="trres" src="https://user-images.githubusercontent.com/10897901/32751879-86a2e3b0-c8c7-11e7-8c73-c754bace4226.png">

ryukinix commented:
> Very nice idea 👍 getting results as comment least it. will avoid going to travis every time to check results stats. well need an account for the organization ! DestructHub Bot ?
( if we will not be obliged to create another docker image xd ) to use PyGithub , or just normal requests with GitHub api ?


We need think a little about that, to do with less effort possible. If you can do that without creating a new user would be better, `codecov` does something similar, so I'm think we can use the travis token or something to accomplish that behavior.

> Btw should we add this feature in this pull request ?

This PullRequest is already big enough and we finish the task well here. We should merge that and open that later since we discuss a great approach about it.

I'm merging your PR now, really thank you to caring about implementing that and all the patience on the lot of my reviews. :) I think we have a nice code for now.

ryukinix commented:
Oh... sorry I only saw now that we forgot to remove the mock/test codes that you added. I should fix that...

A-Hilaly commented:
Sure if codecov did it then we can ! You forgot to remove problem 300/301 ( I already fixed 501)

A-Hilaly commented:
Thank you so much for your time and reviews, I learned so much here and willing for more :) !

ryukinix commented:
> Thank you so much for your time and reviews, I learned so much here and willing for more :) !
Yes, code review is great for everyone. All of us learn something different on code review, as the reviewer and the programmer which has the code reviewed.

> Sure if codecov did it then we can ! You forgot to remove problem 300/301 ( I already fixed 501)

No worries, I'll create a new PR removing that. This even a good thing because we'll saw how this implementation will works when something is deleted from the master (I don't know yet what will happens).

ryukinix commented:
We should move our discussion of comment-build-system-output on #138
